### DISCUSSION TITLE
Loosen deposit amount requirement in `_checkSegments`

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#442] [Ideas]
One of the suggestions left by Rahul Saxena in his audit report was to loosen the deposit amount requirement in `_checkSegments`:

https://github.com/sablierhq/v2-core/blob/main/src/libraries/Helpers.sol#L209

The idea is to handle the user inputs more gracefully by not requiring the deposit amount to be exactly equal to the sum of the segment amounts.

In practice, there are two ways to implement this:

1. Transfer any excess funds to the funder (in the create transaction)
2. Store the excess in the stream struct, and include it in the calcuations in `refundableAmountOf`

The first approach is obviously simpler, since it doesn't break any current protocol invariants and it doesn't interfere with the lifecycle of the stream.

What do you guys think about this proposal? Cc @razgraf, @andreivladbrg.

I'm personally in favor of loosening this requirement, since getting the deposit amount to exactly match the segment amounts sum is not easy. I mean, just look at how convoluted the logic in `Fuzzers.t.sol` is:

https://github.com/sablierhq/v2-core/blob/main/test/utils/Fuzzers.sol

**Comment by PaulRBerg** [comment]
Any thoughts on this proposal, @razgraf, @gavriliumircea?

Have you encountered difficulties in the front-end when calculating the deposit amount for dynamic streams? Would it have helped you if the contract had expected a deposit amount _greater than or equal to_ the segment amounts sum rather than expecting the amounts to be _equal_?

**Comment by gavriliumircea** [comment]
I can't say we had issues at frontend level with this requirement, so from this point of view it does not make much of a difference at front-end level. The important thing IMO is the logic behind the decision. The "natural" behaviour for me is to have the exact deposit when creating a stream, but I understand the other perspective as well and from those 2 options I would choose the first one for its simplicity.(Also I think the first option could be easily associated with getting change when you buy something with cash)

**Reply by PaulRBerg** [reply]
Thanks for your feedback.

> I can't say we had issues at frontend level with this requirement

Good to hear this; however, the reason for having no issues so far is that we haven't enabled protocol and broker fees yet. As shown by [this fuzzing function](https://github.com/sablierhq/v2-core/blob/a8fa096620796395cd08b90a13663427eee18ff9/test/utils/Fuzzers.sol#L99-L116), the calculations get complicated quite quickly once fees are thrown into the equations.

Have you explicitly hard-coded the broker and the protocol fee to zero in the front-end calculations? If not, it would be informative to do so, to remind ourselves of the role that the fees play in determining the create amounts.

> I would choose the first one for its simplicity

Yes, same.

> getting change when you buy something with cash

Great analogy

**Comment by PaulRBerg** [comment]
As per Cantina's feedback on Discord, there are several downsides to loosening this deposit requirement:

- Transferring the excess back to the caller opens up the possibility for reentrancy with tokens that implement hooks
- Storing the excess in the struct would involve a significant refactor
- The fees are calculated on total amount, so the fee paid on the amount that includes the excess

Thus, we will keep the deposit requirement as is. What may be worth it, however, is to provide an on-chain utility for calculating the amounts, but this utility should live in V2 Periphery, not here (in V2 Core).
